### PR TITLE
[MNG-6562] added WARNING on default bindings plugins unlocked versions

### PR DESCRIPTION
**Pull request by hboutemy**

### CODE REVIEW

File: maven-core/src/main/java/org/apache/maven/model/plugin/DefaultLifecycleBindingsInjector.java [hunk 1]
```diff
@@ -87,18 +102,21 @@ else if ( !defaultPlugins.isEmpty() )
     {
 
         private static final String PLUGIN_MANAGEMENT = "plugin-management";
+        private static final String NO_VERSION_PLUGINS = "no-version-plugins";
 
-        public void merge( Model target, Model source )
+        public Collection<Plugin> merge( Model target, Model source )
```
Comment (rfscholte) [inline maven-core/src/main/java/org/apache/maven/model/plugin/DefaultLifecycleBindingsInjector.java]: This looks weird: it says: merge 2 models, the result is a Collection of Plugins.

Comment (hboutemy) [reply maven-core/src/main/java/org/apache/maven/model/plugin/DefaultLifecycleBindingsInjector.java]: true, I forgot to add at least a javadoc to explain the meaning of the return value, which was my minimal idea
reading your comment, I just found another idea: let's pass ModelProblemCollector
I like team work :)

**Comment by hboutemy** [comment]
commit reworked: now passing ModelProblemCollector

File: maven-core/src/main/java/org/apache/maven/model/plugin/DefaultLifecycleBindingsInjector.java [hunk 2]
```diff
@@ -87,18 +88,36 @@ else if ( !defaultPlugins.isEmpty() )
...
 
-            Map<Object, Object> context =
-                Collections.<Object, Object>singletonMap( PLUGIN_MANAGEMENT, target.getBuild().getPluginManagement() );
+            Map<Object, Object> context = new HashMap<Object, Object>();
+            context.put( PLUGIN_MANAGEMENT, target.getBuild().getPluginManagement() );
 
             mergePluginContainer_Plugins( target.getBuild(), source.getBuild(), false, context );
+
+            @SuppressWarnings( "unchecked" )
+            Collection<Plugin> defaultVersionPlugins = (Collection<Plugin>) context.get( NO_VERSION_PLUGINS );
```
Comment (michael-o) [inline maven-core/src/main/java/org/apache/maven/model/plugin/DefaultLifecycleBindingsInjector.java]: That's inconsistent. The field says `defaultVersionPlugins`, but the key `NO_VERSION_PLUGINS`.

File: maven-core/src/main/java/org/apache/maven/model/plugin/DefaultLifecycleBindingsInjector.java [hunk 3]
```diff
@@ -87,18 +88,36 @@ else if ( !defaultPlugins.isEmpty() )
...
+                List<String> plugins = new ArrayList<>( defaultVersionPlugins.size() );
+                for ( Plugin p : defaultVersionPlugins )
+                {
+                    plugins.add( p.getArtifactId() );
+                }
+
+                problems.add( new ModelProblemCollectorRequest( Severity.WARNING, Version.BASE )
+                              .setMessage( "Version not locked for default bindings plugins " + plugins
+                                           + ", you should define versions in pluginManagement section of your "
+                                           + "pom.xml or parent" )
```
Comment (michael-o) [inline maven-core/src/main/java/org/apache/maven/model/plugin/DefaultLifecycleBindingsInjector.java]: that should be consistent too: ... your POM or parent POM. I'd avoid `pom.xml` for polyglot Maven.

**Comment by michael-o** [comment]
@hboutemy Do we want this for 3.6.2?

**Pull request merged**